lib/core/src/client/preview/start.js

```diff
@@ -379,7 +379,10 @@ export default function start(render, { decorateStory } = {}) {
         );
       }
 
-      const { default: meta, ...exports } = fileExports;
+      const { default: meta, __orderedExports, ...namedExports } = fileExports;
+      // prefer a user/loader provided `__orderedExports` object if supplied as es module exports
+      // are ordered alphabetically - see https://github.com/storybookjs/storybook/issues/9136
+      const exports = __orderedExports || namedExports;
```

Comment:
Does js object literal guarantee `Object.keys()` order nowadays? Or would it be safer to use an array instead?

Actually, I think we can use the existing [`іncludeStories`](https://storybook.js.org/docs/formats/component-story-format/#non-story-exports) API for that, instead of introducing a new one

```js
import React from 'react';
import MyComponent from './MyComponent';
import someData from './data.json';

export default {
  title: 'MyComponent',
  component: MyComponent,
  // should also affect the order
  includeStories: ['SimpleStory', 'ComplexStory']
}

export const simpleData = { foo: 1, bar: 'baz' };
export const complexData = { foo: 1, { bar: 'baz', baz: someData }};

export const SimpleStory = () => <MyComponent data={simpleData} />;
export const ComplexStory = () => <MyComponent data={complexData} />;
```

Reply:
Good point @Hypnosphi, tho much harder to implement IMHO

Reply:
as far as I can tell `includeStories` and `__orderedExports` serve a different purpose?

- `includeStories` is handcrafted by users to define specific stories that should be included (and in which order)
- `__orderedExports` should be auto-generated and should include always all stories (can still be filtered with exclude/includeStories

> Does js object literal guarantee Object.keys() order nowadays?

not in the spec but in all practical cases yes. (also the current * import is an Object as well 😬)

We can have an array but it will be more complicated - see https://github.com/storybookjs/storybook/issues/9136#issuecomment-566782481 for an example

imho we should not "misuse"  `includeStories` for it... also there are more questions like what if the users provide includeStories and I want to use it to export the default order? what wins? it would move certain logic into the CSF story files (or its generation/postprocessing) imho not the right place for that logic

Reply:
i agree. it's too complicated to overload `includeStories`!

Reply:
Ok, can it at least share the same API?

```js
export default {
  title: 'MyComponent',
  component: MyComponent,
  storiesOrder: ['SimpleStory', 'ComplexStory']
}
```

Reply:
> much harder to implement

Not really, you just need to iterate over `meta.storiesOrder` instead of `Object.keys(exports)`

Reply:
the complexity is not on the iteration part it is on the loader/processor part

e.g. take this input
```js
export default {
  title: 'MyComponent',
  component: MyComponent
}

export const paragraph = () =>
  html`
    <p>First Story</p>
  `;

export const heading = () =>
  html`
    <h1>Second Story</h1>
  `;
```

now if you want to add the order all you need to do is add on extra line at the end of the file
```js
export __orderedExports = { paragraph, heading };
```

if you want to "inject" it into the export default object then you probably need access to a full AST. 

also imagine something like this

```js
const meta = {};
meta.title = 'MyComponent';

export default meta;
```

or even this
```js
import { meta } from './meta.json';

export default meta;
```

it becomes much harder to inject the data into the right place?

Reply:
Oh, I missed the loader part. How are you going to use a loader without webpack BTW?

Reply:
> is handcrafted by users to define specific stories that should be included (and in which order)

Right now the "in which order" part doesn't work with native ES modules

Reply:
> Oh, I missed the loader part. How are you going to use a loader without webpack BTW?

we are using [es-dev-server](https://open-wc.org/developing/es-dev-server.html) which is based on a koa server (evolvement of express) which can add multiple middlewares where each can process the request before sending it to the client
You can read more about how we use webpack here
https://dev.to/open-wc/storybook-for-web-components-on-steroids-4h29

Also if webpack ever changes the order then storybook will need to add a loader as well